rollup.config.js

```diff
@@ -0,0 +1,252 @@
+import commonjs from 'rollup-plugin-commonjs';
```

Comment:
I find this file extremely complex for its task. IMO a config file should be easy to read and reason about, and this one isn't.

This file tries to avoid code duplication at the expenses of readability and some of its internal functions offer too much flexibility.

I think a more reasonable approach would be to have different files, like `rollup.config.cjs.js`, `rollup.config.esm.js`, `rollup.config.min.js`, `rollup.config.test.js`.

Reply:
> I find this file extremely complex for its task. IMO a config file should be easy to read and reason about, and this one isn't.

It didn't look that complex before I've removed the duplicated items with those functions. 

> I think a more reasonable approach would be to have different files, like...

The complexity would still be the same just spread over multiple files and we would have to add some shared configs which will end up in a similar just distributed complexity. I've had this idea first as well and started to split that up in different module structure type-specific configurations but moved anything back together to one file in the end. I think this is something we also could open an issue about and improve it later.

Reply:
> I find this file extremely complex for its task. IMO a config file should be easy to read and reason about, and this one isn't.

It didn't look that complex before I've removed the duplicated items with those functions. 

> I think a more reasonable approach would be to have different files, like...

The complexity would still be the same just spread over multiple files and we would have to add some shared configs which will end up in a similar just distributed complexity. I've had this idea first as well and started to split that up in different module structure type-specific configurations but moved anything back together to one file in the end. I think this is something we also could open an issue about and improve it later.

Reply:
> some shared configs which will end up in a similar just distributed complexity

Why? I don't follow.

>  I think this is something we also could open an issue about and improve it later.

I disagree with this, as those config files are the core change here, and impact the entire project.

I tried "unrolling" one of the configs, and even that is hard, because there are so much indirection and parameters.

Reply:
I will re-structure the configuration process if this is actually that hard to read. As the creator of this is it really easy to read but it's always another point of view if someone is coming after. :)

Reply:
That would be incredibly helpful. I tried unrolling them (i.e. creating an object with the end result of one config) multiple times, and it's just too much work.

I see the point on having a function and not repeat each thing in every package. I would just use one function for build "type", with as much duplicated config between them as necessary.

So something like:

```js
export function getEsmConfig(...) {
    return {
           // everything inlined in here, without further function calls.
    }
}
```

I guess also using constants, just not so many functions.